servers/src/main/java/tachyon/master/permission/FileSystemPermissionChecker.java

```diff
@@ -49,8 +54,54 @@ public static synchronized void initializeFileSystem(boolean permissionCheckEnab
   }
 
   /**
-   * This method checks access permission.
-   * The input is User and its Groups, requested Permission and inodes (traverse the Path).
+   * Checks requested permission and basic permission on the parent of the path.
+   * Parent means the parent directory of the path.
+   * If parent directory does not exist, treat the closest ancestor directory of the path as
+   * its parent and check permission on it.
+   *
+   * @param user who requests access permission
+   * @param groups in which user belongs to
+   * @param action requested {@link FileSystemAction} by user
+   * @param path whose parent to check permission on
+   * @param fileInfos file infos of all the inodes retrieved by traversing the path
+   * @throws AccessControlException if permission checking fails
+   */
+  public static void checkParentPermission(String user, List<String> groups, FileSystemAction
+      action, TachyonURI path, List<FileInfo> fileInfos) throws AccessControlException,
+      InvalidPathException {
+    String[] pathComponents = PathUtils.getPathComponents(path.getPath());
+
+    if (pathComponents.length == fileInfos.size()) {
+      fileInfos.remove(fileInfos.size() - 1);
+    }
+    checkByFileInfos(user, groups, action, path.getPath(), fileInfos);
+  }
+
+  /**
+   * Checks requested permission and basic permission on the path.
+   *
+   * @param user who requests access permission
+   * @param groups in which user belongs to
+   * @param action requested {@link FileSystemAction} by user
+   * @param path the path to check permission on
+   * @param fileInfos file infos of all the inodes retrieved by traversing the path
+   * @throws AccessControlException if permission checking fails
+   */
+  public static void checkPermission(String user, List<String> groups, FileSystemAction action,
+      TachyonURI path, List<FileInfo> fileInfos) throws AccessControlException,
+      InvalidPathException {
+    String[] pathComponents = PathUtils.getPathComponents(path.getPath());
+
+    for (int i = fileInfos.size(); i < pathComponents.length; i++) {
```

Comment:
`i++` -> `i ++` to match Tachyon code style
http://tachyon-project.org/documentation/Contributing-to-Tachyon.html#coding-style